### ISSUE TITLE
Error in query curve_info for cw20-bonding

### ISSUE THREAD

**Issue by shiki-tak** [CosmWasm/cw-plus#504]
I tried cw20-bonding and got the following error.

```
❯ CURVEINFO='{"curve_info":{}}'
❯ wasmd query wasm contract-state smart $CONTRACT "$CURVEINFO" --output json 
Error: rpc error: code = InvalidArgument desc = Error calling the VM: Error executing Wasm: Wasmer runtime error: RuntimeError: unreachable: query wasm contract failed: invalid request
```
Conditions
- Only query CurveInfo will cause an error.
- Doesn't happen when CurveType is Constant, but happens after Buy for Linear and SquareRoot.
- cw-plus v0.10.0
- wasmd v0.20.0

**Comment by ethanfrey** [comment]
ah, 18 decimals... this causes all kinds of math overflows.

Why do you use that (besides habit from Ethereum)?

I mean, 6 or 9 decimals seems to suffice for almost any project.
The smallest measurement I have seen in use in Ethereum is GigaWei, which effectively is 9 decimal places.
Bitcoin works on 8, and still seems to be fine with that, even with $60.000/BTC (and 16 satoshi to $0.01)

**Comment by maurolacy** [comment]
We can still avoid panic when converting, basically by checking / limiting the range of allowed values / decimal places.

**Comment by shiki-tak** [comment]
I don't have to plan to use 18 decimals.
I think a check or limit to avoid panic is sufficient.

**Comment by ethanfrey** [comment]
I fully agree... a readable error message to the user is *always* better than a panic.

Removing the error condition (precision limit) can be discussed if there is a clear need. We would need to eg use u256 internally for all calculations, which may be a good idea? But checks/limits are 💯 need

**Comment by ethanfrey** [comment]
btw, cw20-bonding was written like 1 year ago and just updated to match versions.
It could probably use some care and cleanup in general.
If you are planning to deploy something on this to production, it would be good to revisit it and feel free to raise issues (and PRs 😄 ) on any points that seem a bit questionable.

I think I have an outstanding issue to allow cw20 as the input tokens, not just native
